### PR TITLE
Fixes for the Avatica JDBC driver

### PR DESCRIPTION
**Pull request by paul-rogers**
PR #12636 amends Druid to make a single pass through the Calcite planner. This was originally done in support of the catalog project to avoid resolving the same catalog object multiple times, which could encounter race conditions if the resolutions occurred on either side of a change to the catalog.

While discussing that change, @clintropolis called our attention to the "query optimized" way that Druid handles parameters, which then raised questions about how we ensure the proposed changes do not break our JDBC support. @gianm suggested we focus testing on JDBC.

That testing revealed that Druid's implementation of the JDBC protocol is a bit off, which may have led to some of the confusion around when to prepare and when to expect parameters. [Issue #12682](https://github.com/apache/druid/issues/12682) and [issue #12684](https://github.com/apache/druid/issues/12684) spell out the gaps.

While JDBC has two kinds of statements (`Statement` and `PreparedStatement`), Druid has just one that attempts to do what the two JDBC statements do. That is, it tries to be both prepared and non-prepared.

Also, while JDBC separates the idea of statement and `ResultSet`, Druid combines the two: closing the result set closes the statement as well. While this might be handy, for a `Statement`, it does work at cross-purposes to the `PreparedStatement` which is designed to be prepared once and executed multiple times. Closing the statement after the first execution forces `PreparedStatement` to be identical to `Statement` (except that the former takes parameters.)

### Corrects the Druid JDBC Driver Behavior

We could argue about the merits of the Druid design. However, JDBC is a standard and the correct behavior is spelled out in the [JDBC 4.1 Spec](https://download.oracle.com/otn-pub/jcp/jdbc-4_1-mrel-spec/jdbc4.1-fr-spec.pdf?AuthParam=1655858260_d6388aeec9aeb8e5616d221749b1ff71). That behavior is:

* A `Statement` is a reusable object that can execute multiple SQL statements, using a `ResultSet`. Closing the `ResultSet` does not close the statement. Druid can, however, close the `ResultSet` on EOF.
* A `PreparedStatement` is a reusable object created with a (typically parameterized) SQL statement. Each execution is given a set of parameter values, and produces a `ResultSet`. Again, closing the `ResultSet` does not close the statement, which allows the statement to be executed with a different set of parameters.

To correct the Druid behavior, the Avatica JDBC implementation is refactored to clearly model the JDBC concepts:

* `DruidJdbcStatement` models a (non-prepared) Statement.
* `DruidJdbcPreparedStatement` models a PreparedStatement.
* `AbstractDruidJdbcStatement` is the common base class for the above.
* `DruidJdbcResultSet` represents a result set (a running query).

While PR #12636 went on to modify the planner and the `SqlLifecyle` layers, this PR leaves those layers unchanged, and implements the desired behavior with the existing code. The hope is that the result is a bit easier on reviewers. This code will be modified again later to support the single-pass planner.

The Avatica driver tests were modified to be consistent with the JDBC standard. The original code didn't pass, but the revised code does.

The result is that we now have a clean SQL stack for JDBC, HTTP and internal use that make a single planner pass when possible, and use two passes for the JDBC prepare/execute model.

### Flaky Test: `DruidAvaticaHandlerTest.testConcurrentQueries()`

Also fixes a flaky test: `DruidAvaticaHandlerTest.testConcurrentQueries()`. This test has occasionally failed in builds. The problem was first mentioned in 2017 in [Issue #4408](https://github.com/apache/druid/issues/4408) , which was claimed to be fixed. When run in a tight loop, however, the test eventually fails with an NPE. The issue is traced to the context on the `DruidConnection`: it is passed to each statement, where it is shared and concurrently modified. The fix is to make a copy for each new statement which avoids queries trying to share the same map.

Verified the problem by running the test 100 times in a tight loop. Verified the fix the same way.

### Minor Revisions

Careful inspection of the code identified several other improvements:

* Hold the yielder thread pool on the connection, not statement, so we get some benefit from caching. (It might be even better to define the pool globally.)
* Define a `SqlRequest` to hold the quad of (SQL, context, parameters, auth result), which makes function signatures simpler.
* Shifted some functionality from `DruidMeta` to `DruidConnection`. `DruidMeta` resolves connections, then connections work with statements.

<hr>

This PR has:

- [X] been self-reviewed.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by paul-rogers** [comment]
Renamed `SqlRequest` to `SqlQueryPlus`. Rebased to resolve merge conflict.

File: sql/src/main/java/org/apache/druid/sql/avatica/DruidMeta.java [hunk 1]
```diff
@@ -229,26 +235,19 @@ public StatementHandle prepare(
...
-      LOG.debug("Successfully prepared statement[%s] for execution", druidStatement.getStatementId());
-      return statement;
+      final DruidConnection druidConnection = getDruidConnection(ch.id);
+      SqlQueryPlus sqlReq = new SqlQueryPlus(
+          sql,
+          null, // Context provided by connection
+          null, // No parameters in this path
+          doAuthenticate(druidConnection)
+      );
+      DruidJdbcPreparedStatement stmt = getDruidConnection(ch.id).createPreparedStatement(
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/avatica/DruidMeta.java:245]: use `druidConnection` instead of calling `getDruidConnection` again?

**Comment by paul-rogers** [comment]
Ready for re-review. The build filed due to a flakey IT recored in [this issue](https://github.com/apache/druid/issues/12795).

File: sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java [hunk 2]
```diff
@@ -89,36 +107,85 @@ public DruidStatement createStatement(SqlLifecycleFactory sqlLifecycleFactory)
...
+        // Will only happen if statementCounter rolls over before old statements are cleaned up. If this
+        // ever happens then something fishy is going on, because we shouldn't have billions of statements.
+        throw DruidMeta.logFailure(new ISE("Uh oh, too many statements"));
+      }
+
+      if (statements.size() >= maxStatements) {
+        throw DruidMeta.logFailure(new ISE("Too many open statements, limit is [%,d]", maxStatements));
+      }
+
+      @SuppressWarnings("GuardedBy")
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java:144]: does this still need suppressed?

**Review by clintropolis — APPROVED** [review]
:+1:

The only thing i'm unsure about is if the executor change for running stuff changes the characteristics of anything, but not sure how to find out without merging it to see

**Comment by paul-rogers** [comment]
Release note item:

In previous releases, Druid would automatically close the JDBC `Statement` when the `ResultSet` is closed. Druid would close the `ResultSet` on EOF. Druid would close the statement on any exception. This behavior is, however, non-standard.

In this release, Druid's JDBC driver follows the JDBC standards more closely:

* The `ResultSet` closes automatically on EOF, but does not close the `Statement` or `PreparedStatement`. Your code must close these statements, perhaps by using a try-with-resources block.
* The `PreparedStatement` can now be used multiple times with different parameters. (Previously this was not true since closing the `ResultSet` closed the `PreparedStatement`.)
* If any call to a `Statement` or `PreparedStatement` raises an error, the client code must still explicitly close the statement. According to the JDBC standards, statements are not closed automatically on errors. This allows you to obtain information about a failed statement before closing it.

If you have code that depended on the old behavior, you may have to change your code to add the required close statement.

**Pull request merged**